testing_test.go

```diff
@@ -0,0 +1,287 @@
+	runner := &fakeRunner{}
+	hyp := `error {"msg":"There was a problem.","data":"3.14"}
+debug {"msg":"Here's the number of calls.","numCalls":"18"}
+`
+
+	var logBuf bytes.Buffer
+	logger := NewTestingLogger(runner, WithTruthString(hyp), WithLogWriter(&logBuf))
+
+	logger.Error("msg", "There was a problem.", "data", 3.14)
+	logger.Debug("msg", "Here's some pertinent information.", "numCalls", 18) // wrong!
+	logger.Trace("msg", "This trace message shouldn't be here.")
+
+	logger.Done()
+
+	fmt.Print(runner.b.String())
+	fmt.Print(logBuf.String())
+	fmt.Println(runner.failed)
+	// Output:
+	// unexpected log message (-want +got):
+	//   string(
+	// - 	`debug {"msg":"Here's the number of calls.","numCalls":"18"}`,
+	// + 	`debug {"msg":"Here's some pertinent information.","numCalls":"18"}`,
+	//   )
+	// unexpected log message (-want +got):
+	//   string(
+	// - 	"",
+	// + 	`trace {"msg":"This trace message shouldn't be here."}`,
+	//   )
+	// error {"msg":"There was a problem.","data":"3.14"}
+	// debug {"msg":"Here's some pertinent information.","numCalls":"18"}
+	// trace {"msg":"This trace message shouldn't be here."}
+	// true
+}
+
+func ExampleWithoutFailure() {
+	runner := &fakeRunner{}
+	hyp := `error {"msg":"There was a problem.","data":"3.14"}
+debug {"msg":"Here's the number of calls.","numCalls":"19"}
+`
+	logger := NewTestingLogger(runner, WithTruthString(hyp), WithoutFailure())
+
+	logger.Error("msg", "There was a problem.", "data", 3.14)
+
+	// This log message doesn't match, but we still want the test to pass.
+	logger.Debug("msg", "Here's some pertinent information.", "numCalls", 19)
+
+	logger.Done()
+
+	fmt.Println(runner.failed)
+	// Output:
+	// false
+}
+
+type testingLogMsg struct {
+	lvl     level.Level
+	keyvals []interface{}
+}
+
+func TestWithTruthString(t *testing.T) { // nolint: funlen
+	tests := map[string]struct {
+		// in is the set of log messages that will actually be given to the TestingLogger.
+		in []testingLogMsg
+
+		// expect is the set of log messages that the TestingLogger should expect.
+		expect string
+
+		// hyp is the complete string that we should expect to be written to the test runner given
+		// `expect` and `in`.
+		hyp string
+
+		// expectFail is whether to expect the TestingLogger to call runner.Fail.
+		expectFail bool
+	}{
+		"empty": {
+			expectFail: false,
+		},
+		"correct": {
+			in: []testingLogMsg{
+				{level.Trace, []interface{}{"msg", "This is just a trace.", "data", 3.14}},
+				{level.Debug, []interface{}{"msg", "This is a debug msg.", "data", []byte{0, 1, 2, 3}}},
+				{level.Info, []interface{}{"msg", "This msg might be useful.", "data", 12}},
+				{level.Error, []interface{}{"msg", "There's a problem.", "data"}}, // missing value
+			},
+			expect: `trace {"msg":"This is just a trace.","data":"3.14"}
+debug {"msg":"This is a debug msg.","data":"[0 1 2 3]"}
+info  {"msg":"This msg might be useful.","data":"12"}
+error {"msg":"There's a problem.","data":"missing"}`,
```

Comment:
Would it be more readable if we did a strings.Join?  Shoot me down if you don't like it.  It propbably wouldn't resolve anything very well.
```suggestion
			expect: strings.Join([]string{
				`trace {"msg":"This is just a trace.","data":"3.14"}`,
				`debug {"msg":"This is a debug msg.","data":"[0 1 2 3]"}`,
				`info  {"msg":"This msg might be useful.","data":"12"}`,
				`error {"msg":"There's a problem.","data":"missing"}`,
			}, "\n")
```

Reply:
good idea